### DISCUSSION TITLE
Will we support Object JSON serialize?

### DISCUSSION THREAD

**Discussion by SunFulong** [wamp-proto/wamp-proto#481] [Q&A] [answered]
First of all, I know a JSON text can be object or ARRAY, and we choose ARRAY way.

But, when I trying to develop with AWS WebSocket API, AWS only support explain property of OBJECT as route key.

So, will we support the OBJECT way serialize?

**Comment by oberstet** [comment]
I don't understand the question: every single element of `args` can be an object, the whole `kwargs` must be an object. So just serialize your object and put it into the right item ...

**Comment by KSDaemon** [comment, marked as answer]
@SunFulong If you mean that the whole WAMP message can be an object instead of an array as it is now — sorry, noway.

**Reply by SunFulong** [reply]
OK, thank you for your firm answer, I should make a WAMP-like protocol for AWS.

**Comment by oberstet** [comment]
yes, true. it wouldn't make sense (it would be a different protocol), and it is not needed

**Reply by SunFulong** [reply]
The only sense is: it will be compatible with AWS WebSocket API Gateway.

**Reply by oberstet** [reply]
Or, you could use a proper, more powerful and likely cheaper WebSocket gateway like Cloudflare.

**Reply by SunFulong** [reply]
> Or, you could use a proper, more powerful and likely cheaper WebSocket gateway like Cloudflare.

Could you give me a link? I didn't google it out

**Comment by ecorm** [comment]
If you're going to adopt the AWS ecosystem for your project, they may already have something similar to WAMP you could adopt. I am not familiar with the AWS tech stack.

If the idea of adopting WAMP is not to be tied to a vendor's proprietary tech stack, perhaps you could add an intermediary layer that wraps the WAMP array message into an object?

**Reply by SunFulong** [reply]
> perhaps you could add an intermediary layer that wraps the WAMP array message into an object?

Yes, I will wrap the ARRAY-JSON message to OBJECT-JSON message, for each element add the NAME key describes in spec, e.g. `{message_type: 1, realm: "somerealm"}`

> If you're going to adopt the AWS ecosystem for your project

No, not ecosystem, only their SAM (Amazon Serverless Application Model). I put a root WAMP router on SAM, and connect our IDCs via WAMP (like) protocol, each IDC have an edge WAMP router and multiple business clients.

**Reply by ecorm** [reply]
I believe this wrapping approach would be called the _Envelope Wrapper_ pattern, if you're looking for a name: https://www.enterpriseintegrationpatterns.com/patterns/messaging/EnvelopeWrapper.html

**Reply by SunFulong** [reply]
Looks like, but not, because I didn't unwrap the message back to raw, instead, process the wrapped message directly.

**Reply by oberstet** [reply]
However you call the resulting mess, please don't call it WAMP ;)

**Reply by SunFulong** [reply]
> However you call the resulting mess, please don't call it WAMP ;)

Yes, I won't call it WAMP, but will in my document write "foo protocol - a WAMP like protocol"
